share/github-backup-utils/ghe-restore-repositories

```diff
@@ -0,0 +1,177 @@
+
+# Disable remote GC operations
+for hostname in $hostnames; do
+  ghe-gc-disable $ssh_config_file_opt $hostname:$port
+done
+
+# Build a list of network paths to send to the server to calculate
+# the restore routes, something like:
+#
+# a/nw/a8/3f/02/100000855
+# a/nw/a8/bc/8d/100000880
+# a/nw/a5/06/81/100000659
+# a/nw/a5/84/6f/100000708
+# a/nw/a5/e0/01/146
+# ...
+#
+# One network path per line.
+bm_start "$(basename $0) - Building network list"
+OLDIFS=$IFS; IFS=$'\n'
+for path in $network_paths; do
+   # Get the network ID
+   # The network id from a repository is the last component of the path
+   # i.e. /data/repositories/a/nw/a5/bf/c9/37 network ID would be 37
+   ghe_verbose "* Adding network_path $path to the list of networks to send"
+   echo $path
+done > $tmp_list
+IFS=$OLDIFS
+bm_end "$(basename $0) - Building network list"
+
+# The server returns a list of routes:
+#
+# a/nw/a8/3f/02/100000855 dgit-node1 dgit-node2 dgit-node3
+# a/nw/a8/bc/8d/100000880 dgit-node1 dgit-node2 dgit-node4
+# a/nw/a5/06/81/100000659 dgit-node3 dgit-node2 dgit-node4
+# ...
+#
+# One route per line.
+#
+bm_start "$(basename $0) - Calculating sync routes"
+cat $tmp_list | ghe-ssh "$GHE_HOSTNAME" -- github-env ./bin/dgit-cluster-restore-routes \
+  | while read route; do
+  ghe_debug "Got restore route $route"
+  if $CLUSTER; then
+    servers=$(echo $route | cut -d ' ' -f2-)
+  else
+    servers=$host
+  fi
+  network_path=$(echo $route | cut -d ' ' -f1)
+  for server in $servers; do
+    ghe_debug "Adding $network_path to $tempdir/$server.rsync"
+    echo "$network_path" >> $tempdir/$server.rsync
+    echo "$route" | awk '{ n = split($1, p, "/"); printf p[n] " /data/repositories/" $1; $1=""; print $0}' >> $to_restore
+  done
+done
+bm_end "$(basename $0) - Calculating sync routes"
+
+if ! ls $tempdir/*.rsync >/dev/null 2>&1; then
+  echo "Warning: no routes found, skipping repositories restore ..."
+  exit 0
+fi
+
+bm_start "$(basename $0) - Restoring repository networks"
+# rsync all the repositories
+for file_list in $tempdir/*.rsync; do
+  server=$(basename $file_list .rsync)
+  ghe_verbose "* Transferring repository networks to $server ..."
+  ghe-rsync -avrHR --delete \
+    -e "ssh -q $opts -p $port $ssh_config_file_opt -l $user" \
+    --rsync-path="sudo -u git rsync" \
+    --files-from=$file_list \
+    "$GHE_DATA_DIR/$GHE_RESTORE_SNAPSHOT/repositories/./" \
+    "$server:$GHE_REMOTE_DATA_USER_DIR/repositories/" 1>&3
+done
+bm_end "$(basename $0) - Restoring repository networks"
+
+# Tell dgit about the repositories restored
+if $CLUSTER; then
+  bm_start "$(basename $0) - Finalizing routes"
+  ghe_verbose "Finalizing routes"
+  cat $to_restore | ghe-ssh "$GHE_HOSTNAME" -- sponge $to_restore
+  ghe-ssh "$GHE_HOSTNAME" -- /bin/bash >&3 <<EOF
+    split -l 1000 $to_restore $tempdir/chunk
+    chunks=\$(find $tempdir/ -name chunk\*)
+    echo "\$chunks" | xargs -P64 -n1 -I{} /bin/sh -c "cat {} | github-env ./bin/dgit-cluster-restore-finalize"
```

Comment:
Heads up @snh. Since you last looked at this, I've switched out the use of `parallel` in favour of `xargs -P64` for all of the finalize steps (with a few force pushes to hide my mess 😊). 

My testing shows this works just as well, but I'd appreciate it if you could specifically double check me here - I'm not sure why you went for `parallel` over `xargs` when you first implemented this. 🙇

Reply:
@lildude I'm concerned that 64 will be a little high in some circumstances, as there are a number of Cluster customers who restore to DR and test clusters that are quite light-on resources wise.

From my quick testing, each `dgit-cluster-restore-finalize`, without having been feed any routes yet, uses 76MB of PSS memory, which adds up to close to 5GB for 64 of these.

For this reason I'm quite a fan of the approach we were using with `parallel` to scale the number of processes based on the CPU count, which is often a good indicator of the resources a node has available to it.

Did you have a chance to try and stub `parallel`? As this runs on the appliance itself, which already had `parallel`, the dependency on `parallel` has never really been an issue until now.

Reply:
> I'm concerned that 64 will be a little high in some circumstances, as there are a number of Cluster customers who restore to DR and test clusters that are quite light-on resources wise.

This would only affect customers who actually have up to 64 CPUs on their node as `-P` is for maximum number of processes. From the man page (emphasis is mine): "Run  **up  to**  max-procs processes at a time;". 

I'd hope if they're going to assign that many CPUs, they'd allocate an equivalently large amount of memory too as they'd hit other problems long before they hit backup-utils related problems.

> For this reason I'm quite a fan of the approach we were using with parallel to scale the number of processes based on the CPU count

I'd have liked to do the same with `xargs` by not specifying a number or using 0 to use all available procs, but `xargs` on macOS requires a value > 0 and we still test on macOS, hence I've gone for a rather large value to mimic the same behaviour.

> Did you have a chance to try and stub parallel? 

I did, but this turned out to be a lot more complex than anticipated and I soon found myself effectively writing a passthrough version of `parallel` that stripped the args so we could run the subsequent commands. This adds unnecessary complexity and potential tech debt, just for testing, when a simple and just as effective alternate solution works for both actual use and testing.

> As this runs on the appliance itself, which already had parallel, the dependency on parallel has never really been an issue until now.

That's because we were never testing this behaviour within backup-utils itself so never encountered any problems until now. We don't use `parallel` for anything else within GitHub Enterprise either.

Reply:
As discussed in Slack, I'm wrong here and misunderstood the behaviour. Will revert that change.